twilight-gateway/src/json.rs

```diff
@@ -0,0 +1,160 @@
+#[derive(Debug)]
+pub struct GatewayEventParsingError {
+    /// Type of error.
+    pub(crate) kind: GatewayEventParsingErrorType,
+    /// Source error if available.
+    pub(crate) source: Option<Box<dyn Error + Send + Sync>>,
+}
+
+impl GatewayEventParsingError {
+    /// Shortcut to create a new error from an invalid UTF-8 error.
+    fn from_utf8(source: Utf8Error) -> Self {
+        Self {
+            kind: GatewayEventParsingErrorType::PayloadInvalid,
+            source: Some(Box::new(source)),
+        }
+    }
+}
+
+impl Display for GatewayEventParsingError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        match self.kind {
+            GatewayEventParsingErrorType::Deserializing => {
+                f.write_str("deserializing gateway event as json failed")
+            }
+            GatewayEventParsingErrorType::PayloadInvalid => {
+                f.write_str("payload is an invalid json structure")
+            }
+        }
+    }
+}
+
+impl Error for GatewayEventParsingError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        self.source
+            .as_ref()
+            .map(|source| &**source as &(dyn Error + 'static))
+    }
+}
+
+/// Type of [`GatewayEventParsingError`] that occurred.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum GatewayEventParsingErrorType {
+    /// Deserializing the GatewayEvent payload from JSON failed.
+    Deserializing,
+    /// The payload received from Discord was an unrecognized or invalid
+    /// structure.
+    ///
+    /// The payload was either invalid JSON, not UTF-8 valid, or did not contain
+    /// the necessary "op" key in the object.
+    PayloadInvalid,
+}
+
+/// Wrapper over [`from_slice_inner`], wrapping the error type with a custom
+/// one.
+///
+/// # Errors
+///
+/// Returns a [`GatewayEventParsingErrorType::Deserializing`] error type if the
+/// payload failed to deserialize.
+pub fn from_slice<T: DeserializeOwned>(json: &mut [u8]) -> Result<T, GatewayEventParsingError> {
+    from_slice_inner(json).map_err(|source| {
+        tracing::error!("invalid JSON: {}", String::from_utf8_lossy(json));
+
+        GatewayEventParsingError {
+            kind: GatewayEventParsingErrorType::Deserializing,
+            source: Some(Box::new(source)),
+        }
+    })
+}
+
+/// Parse JSON into a gateway event without existing knowledge of its underlying
+/// parts.
+///
+/// # Errors
+///
+/// Returns a [`GatewayEventParsingErrorType::Deserializing`] error type if the
+/// payload failed to deserialize.
+///
+/// Returns a [`GatewayEventParsingErrorType::PayloadInvalid`] error type if the
+/// payload wasn't a valid `GatewayEvent` data structure, such as due to not
+/// being UTF-8 valid.
+pub fn parse(json: &mut [u8]) -> Result<GatewayEvent, GatewayEventParsingError> {
+    #[cfg(not(feature = "simd-json"))]
+    let text = str::from_utf8(json).map_err(GatewayEventParsingError::from_utf8)?;
```

Comment:
`text` is needed for the duration of `serde_json`'s deserialization logic as `GatewayEventDeserializer` takes a reference to it, but `simd-json` uses `GatewayEventDeserializeOwned` because simd-json needs a mutable reference to the buffer. We can't have both an immutable and mutable reference, so we can't do that.